packages/loader/container-loader/src/container.ts

```diff
@@ -379,81 +379,104 @@ export class Container
-								// the telemetry where users sometimes search for ContainerClose event to look
-								// for load failures. So not removing this at this time.
-								container.close(err);
-								container.dispose(err);
-								onClosed(err);
-							},
-						);
-				}),
-			{ start: true, end: true, cancel: "generic" },
-		);
+		const initialize = async (): Promise<void> =>
+			PerformanceEvent.timedExecAsync(
+				container.mc.logger,
+				{ eventName: "Load", ...loadMode },
+				async (event) =>
+					new Promise((resolve, reject) => {
+						const defaultMode: IContainerLoadMode = { opsBeforeReturn: "cached" };
+						// if we have pendingLocalState, anything we cached is not useful and we shouldn't wait for connection
+						// to return container, so ignore this value and use undefined for opsBeforeReturn
+						const mode: IContainerLoadMode = pendingLocalState
+							? { ...(loadMode ?? defaultMode), opsBeforeReturn: undefined }
+							: (loadMode ?? defaultMode);
+
+						const onClosed = (err?: ICriticalContainerError): void => {
+							// pre-0.58 error message: containerClosedWithoutErrorDuringLoad
+							reject(err ?? new GenericError("Container closed without error during load"));
+						};
+						container.on("closed", onClosed);
+
+						container
+							.load(version, mode, resolvedUrl, pendingLocalState)
+							.finally(() => {
+								container.removeListener("closed", onClosed);
+							})
+							.then(
+								(props) => {
+									event.end({ ...props });
+									resolve();
+								},
+								(error) => {
+									const err = normalizeError(error);
+									// Depending where error happens, we can be attempting to connect to web socket
+									// and continuously retrying (consider offline mode)
+									// Host has no container to close, so it's prudent to do it here
+									// Note: We could only dispose the container instead of just close but that would
+									// the telemetry where users sometimes search for ContainerClose event to look
+									// for load failures. So not removing this at this time.
+									container.close(err);
+									container.dispose(err);
+									onClosed(err);
+								},
+							);
+					}),
+				{ start: true, end: true, cancel: "generic" },
+			);
+
+		return Object.assign(container, { initialize });
 	}
 
 	/**
 	 * Create a new container in a detached state.
 	 */
-	public static async createDetached(
+	public static createDetached(
 		createProps: IContainerCreateProps,
 		codeDetails: IFluidCodeDetails,
-	): Promise<Container> {
+	): Container & { initialize: () => Promise<void> } {
 		const container = new Container(createProps);
 
-		return PerformanceEvent.timedExecAsync(
-			container.mc.logger,
-			{ eventName: "CreateDetached" },
-			async (_event) => {
-				await container.createDetached(codeDetails);
-				return container;
-			},
-			{ start: true, end: true, cancel: "generic" },
-		);
+		return Object.assign(container, {
```

Comment:
If I got it right, you want symmetrical flow, i.e. call initialize() in both cases.
I'm not sure I like it - while symmetry is nice, it exposes container to users in a new state, which means we now need to test all the things that user can do with such container in this state (before initialize() is called), even if the message is - do not call any API. I'd prefer the flow I described above, and leave detached flow as is, i.e. single call returns detached container.